pkg/prometheus/operator.go

```diff
@@ -1763,39 +1763,75 @@ func (c *Operator) selectServiceMonitors(ctx context.Context, p *monitoringv1.Pr
 	res := make(map[string]*monitoringv1.ServiceMonitor, len(serviceMonitors))
 	for namespaceAndName, sm := range serviceMonitors {
 		var err error
-
+		rejectFn := func(probe *monitoringv1.ServiceMonitor, err error) {
+			rejected++
+			level.Warn(c.logger).Log(
+				"msg", "skipping servicemonitor",
+				"error", err.Error(),
+				"servicemonitor", namespaceAndName,
+				"namespace", p.Namespace,
+				"prometheus", p.Name,
+			)
+		}
 		for i, endpoint := range sm.Spec.Endpoints {
 			// If denied by Prometheus spec, filter out all service monitors that access
 			// the file system.
 			if p.Spec.ArbitraryFSAccessThroughSMs.Deny {
 				if err = testForArbitraryFSAccess(endpoint); err != nil {
-					break
+					rejectFn(sm, err)
+					continue
 				}
 			}
 
 			smKey := fmt.Sprintf("serviceMonitor/%s/%s/%d", sm.GetNamespace(), sm.GetName(), i)
 
 			if err = store.AddBearerToken(ctx, sm.GetNamespace(), endpoint.BearerTokenSecret, smKey); err != nil {
-				break
+				rejectFn(sm, err)
+				continue
 			}
 
 			if err = store.AddBasicAuth(ctx, sm.GetNamespace(), endpoint.BasicAuth, smKey); err != nil {
-				break
+				rejectFn(sm, err)
+				continue
 			}
 
 			if endpoint.TLSConfig != nil {
 				if err = store.AddTLSConfig(ctx, sm.GetNamespace(), endpoint.TLSConfig); err != nil {
-					break
+					rejectFn(sm, err)
+					continue
 				}
 			}
 
 			if err = store.AddOAuth2(ctx, sm.GetNamespace(), endpoint.OAuth2, smKey); err != nil {
-				break
+				rejectFn(sm, err)
+				continue
 			}
 
 			smAuthKey := fmt.Sprintf("serviceMonitor/auth/%s/%s/%d", sm.GetNamespace(), sm.GetName(), i)
 			if err = store.AddSafeAuthorizationCredentials(ctx, sm.GetNamespace(), endpoint.Authorization, smAuthKey); err != nil {
-				break
+				rejectFn(sm, err)
+				continue
```

Comment:
For this and the similar changes above, I am wondering how they relate to validation of scrape timeout? Should they be dealt with in a different PR?

Also wondering what drove this change? Why do we continue to process the rest of the service monitor endpoints in case of error as opposed to dropping the whole CR? Wouldn't the log `skipping servicemonitor` be misleading since we may end up with a "partially reconciled" ServiceMonitor?

Reply:
I agree with @PhilipGough: we should drop the whole service monitor (same for pod monitors).

Reply:
@PhilipGough  Sorry I was still updating the PR :) 

I was trying to make it like as `selectProbes`  because I wanted to wrap the error returned from validate function.

https://github.com/prometheus-operator/prometheus-operator/blob/14b5ca17052ea16b22bacee31354d46de4430ab4/pkg/prometheus/operator.go#L2004-L2056

I think we would need to update `selectProbe` too then?